### ISSUE TITLE
npm install error

### ISSUE THREAD

**Issue by FerrariAndrea** [rla/node-swipl#7]
$ npm install swipl

> swipl@1.0.3 install C:\*****\node_modules\swipl
> node-gyp rebuild

C:\*****\node_modules\swipl>if not defined npm_config_node_gyp (node "C:\Program Files\nodejs\node_modules\npm\node_modules\npm-lifecycle\node-gyp-bin\\..\..\node_modules\node-gyp\bin\node-gyp.js" rebuild )  else (node "C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\bin\node-gyp.js" rebuild )
gyp ERR! configure error
gyp ERR! stack Error: Command failed: C:\Users\*****\AppData\Local\Programs\Python\Python37-32\python.EXE -c import sys; print "%s.%s.%s" % sys.version_info[:3];
gyp ERR! stack   File "<string>", line 1
gyp ERR! stack     import sys; print "%s.%s.%s" % sys.version_info[:3];
gyp ERR! stack                                ^
gyp ERR! stack SyntaxError: invalid syntax
gyp ERR! stack
gyp ERR! stack     at ChildProcess.exithandler (child_process.js:294:12)
gyp ERR! stack     at ChildProcess.emit (events.js:189:13)
gyp ERR! stack     at maybeClose (internal/child_process.js:970:16)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:259:5)
gyp ERR! System Windows_NT 10.0.17134
gyp ERR! command "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\node_modules\\node-gyp\\bin\\node-gyp.js" "rebuild"
gyp ERR! cwd C:*****\node_modules\swipl
gyp ERR! node -v v10.15.2
gyp ERR! node-gyp -v v3.8.0
gyp ERR! not ok
npm WARN compilation_helper@1.0.0 No description
npm WARN compilation_helper@1.0.0 No repository field.

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! swipl@1.0.3 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the swipl@1.0.3 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\*****\AppData\Roaming\npm-cache\_logs\2019-10-22T11_19_44_040Z-debug.log

**Comment by rla** [comment]
Node-gyp version in package.json is probably too old. It should be easy to upgrade. I will do it in the couple of next days.

**Comment by FerrariAndrea** [comment]
thx, i don't know if it is related to this issues, I noticed that on the npm install --global --production windows-build-tools (as Admin user) the process don't end, it remains stuck as this:


```
C:\MyData\CompilationHelper\Server>npm install --global --production windows-build-tools

> windows-build-tools@5.2.2 postinstall C:\Users\*****\AppData\Roaming\npm\node_modules\windows-build-tools
> node ./dist/index.js



Downloading python-2.7.15.amd64.msi
[>                                            ] 0.0% (0 B/s)
Downloaded python-2.7.15.amd64.msi. Saved to C:\Users\*****\.windows-build-tools\python-2.7.15.amd64.msi.
Downloading vs_BuildTools.exe
[>                                            ] 0.0% (0 B/s)
Downloaded vs_BuildTools.exe. Saved to C:\Users\*****\.windows-build-tools\vs_BuildTools.exe.

Starting installation...
Launched installers, now waiting for them to finish.
This will likely take some time - please be patient!

Status from the installers:
---------- Visual Studio Build Tools ----------
2019-10-22T13:51:23 : Verbose : Visual Studio Installer (1.18.1104.625 : release) ["C:\\Program Files (x86)\\Microsoft Visual Studio\\Installer\\vs_installershell.exe","/finalizeInstall","install","--in","C:\\ProgramData\\Microsoft\\VisualStudio\\Packages\\_bootstrapper\\vs_setup_bootstrapper_201910221351209102.json","--norestart","--quiet","--includeRecommended","--add","Microsoft.VisualStudio.Workload.VCTools","--locale","it-IT","--activityId","f368a1d4-c48c-4193-b88f-40d9bd4ef90b"]
2019-10-22T13:51:24 : Verbose : Installation finalized successfully.
------------------- Python --------------------
---------- Visual Studio Build Tools ----------
2019-10-22T13:51:23 : Verbose : Visual Studio Installer (1.18.1104.625 : release) ["C:\\Program Files (x86)\\Microsoft Visual Studio\\Installer\\vs_installershell.exe","/finalizeInstall","install","--in","C:\\ProgramData\\Microsoft\\VisualStudio\\Packages\\_bootstrapper\\vs_setup_bootstrapper_201910221351209102.json","--norestart","--quiet","--includeRecommended","--add","Microsoft.VisualStudio.Workload.VCTools","--locale","it-IT","--activityId","f368a1d4-c48c-4193-b88f-40d9bd4ef90b"]
2019-10-22T13:51:24 : Verbose : Installation finalized successfully.
------------------- Python --------------------
Successfully installed Python 2.7

```

it seems he has to do something else

**Comment by rla** [comment]
You should check if it's doing something in the background (from Task Manager).

**Comment by FerrariAndrea** [comment]
for the "npm install --global --production windows-build-tools" was the wifii which restrich the update. 

For npm install swipl the error persist:
```

C:\***\Server>npm i swipl

> swipl@1.0.3 install C:\***\Server\node_modules\swipl
> node-gyp rebuild


C:\***\Server\node_modules\swipl>if not defined npm_config_node_gyp (node "C:\Program Files\nodejs\node_modules\npm\node_modules\npm-lifecycle\node-gyp-bin\\..\..\node_modules\node-gyp\bin\node-gyp.js" rebuild )  else (node "C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\bin\node-gyp.js" rebuild )
"swipl" non è riconosciuto come comando interno o esterno,
 un programma eseguibile o un file batch.
child_process.js:656
    throw err;
    ^

Error: Command failed: swipl --dump-runtime-variables=sh
"swipl" non ´┐¢ riconosciuto come comando interno o esterno,
 un programma eseguibile o un file batch.

    at checkExecSyncError (child_process.js:616:11)
    at execSync (child_process.js:653:13)
    at Object.exports.dump (C:\***\Server\node_modules\swipl\lib\runtime_vars.js:7:20)
    at Object.<anonymous> (C:\***\Server\node_modules\swipl\scripts\runtime_var.js:10:26)
    at Module._compile (internal/modules/cjs/loader.js:689:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:700:10)
    at Module.load (internal/modules/cjs/loader.js:599:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:538:12)
    at Function.Module._load (internal/modules/cjs/loader.js:530:3)
    at Function.Module.runMain (internal/modules/cjs/loader.js:742:12)
gyp: Call to 'node scripts/runtime_var.js PLARCH' returned exit status 1 while in binding.gyp. while trying to load binding.gyp
gyp ERR! configure error
gyp ERR! stack Error: `gyp` failed with exit code: 1
gyp ERR! stack     at ChildProcess.onCpExit (C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\lib\configure.js:345:16)
gyp ERR! stack     at ChildProcess.emit (events.js:189:13)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:248:12)
gyp ERR! System Windows_NT 10.0.17134
gyp ERR! command "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\node_modules\\node-gyp\\bin\\node-gyp.js" "rebuild"
gyp ERR! cwd C:\MyData\CompilationHelper\Server\node_modules\swipl
gyp ERR! node -v v10.15.2
gyp ERR! node-gyp -v v3.8.0
gyp ERR! not ok
npm WARN compilation_helper@1.0.0 No description
npm WARN compilation_helper@1.0.0 No repository field.

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! swipl@1.0.3 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the swipl@1.0.3 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\***\AppData\Roaming\npm-cache\_logs\2019-10-23T05_50_24_939Z-debug.log
```

Then i try this: npm install node-gyp rebuild
```
+ node-gyp@6.0.0
+ rebuild@0.1.2
added 95 packages from 62 contributors and audited 318 packages in 13.831s
found 0 vulnerabilities
```

but the result does not change

Edit:
i try to download the repo, and do "npm install" on 2 different pc, the error is the same

**Comment by rla** [comment]
@FerrariAndrea you need to have SWI-Prolog installed and `swipl` command must be available in PATH. This error is indicating that `swipl` is not available in PATH:

```
"swipl" non ´┐¢ riconosciuto come comando interno o esterno,
 un programma eseguibile o un file batch.
```

Yesterday I updated repo to include node-gyp 6 as dep but I haven't published new version of the package yet. I will try to do this today. Anyway, you still need to install SWI first.

**Comment by FerrariAndrea** [comment]
Yes, you are right, the installation of SWI-Prolog don't add
automatically the `swipl ` on the enviroment variables.
Now swipl cmd work fine and on npm install i get this:

```
C:\Users\*****\Desktop\node-swipl-master>npm install

> swipl@1.0.4 install C:\Users\*****\Desktop\node-swipl-master
> node-gyp rebuild


C:\Users\*****\Desktop\node-swipl-master>if not defined npm_config_node_gyp (node "C:\Program Files\nodejs\node_modules\npm\node_modules\npm-lifecycle\node-gyp-bin\\..\..\node_modules\node-gyp\bin\node-gyp.js" rebuild )  else (node "C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\bin\node-gyp.js" rebuild )
Compilazione dei progetti nella soluzione uno alla volta. Per abilitare la compilazione parallela, aggiungere l'opzione "/m".
  libswipl.cc
  win_delay_load_hook.cc
c:\users\*****\desktop\node-swipl-master\src\libswipl.cc(77): error C3861: 'PL_get_compound_name_arity': ident
ificatore non trovato [C:\Users\*****\Desktop\node-swipl-master\build\swipl.vcxproj]
c:\users\*****\desktop\node-swipl-master\src\libswipl.cc(167): error C2065: 'PL_NIL': identificatore non dichi
arato [C:\Users\*****\Desktop\node-swipl-master\build\swipl.vcxproj]
c:\users\*****\desktop\node-swipl-master\src\libswipl.cc(169): error C2065: 'PL_LIST_PAIR': identificatore non
 dichiarato [C:\Users\*****\Desktop\node-swipl-master\build\swipl.vcxproj]
c:\users\*****\desktop\node-swipl-master\src\libswipl.cc(179): error C2065: 'PL_BLOB': identificatore non dich
iarato [C:\Users\*****\Desktop\node-swipl-master\build\swipl.vcxproj]
c:\users\*****\desktop\node-swipl-master\src\libswipl.cc(182): error C2065: 'PL_DICT': identificatore non dich
iarato [C:\Users\*****\Desktop\node-swipl-master\build\swipl.vcxproj]
c:\users\*****\desktop\node-swipl-master\src\libswipl.cc(167): error C2051: espressione case non costante [C:\
Users\*****\Desktop\node-swipl-master\build\swipl.vcxproj]
c:\users\*****\desktop\node-swipl-master\src\libswipl.cc(169): error C2051: espressione case non costante [C:\
Users\*****\Desktop\node-swipl-master\build\swipl.vcxproj]
c:\users\*****\desktop\node-swipl-master\src\libswipl.cc(179): error C2051: espressione case non costante [C:\
Users\*****\Desktop\node-swipl-master\build\swipl.vcxproj]
c:\users\*****\desktop\node-swipl-master\src\libswipl.cc(182): error C2051: espressione case non costante [C:\
Users\*****\Desktop\node-swipl-master\build\swipl.vcxproj]
gyp ERR! build error
gyp ERR! stack Error: `C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\MSBuild\15.0\Bin\MSBuild.exe` failed with exit code: 1
gyp ERR! stack     at ChildProcess.onExit (C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\lib\build.js:262:23)
gyp ERR! stack     at ChildProcess.emit (events.js:189:13)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:248:12)
gyp ERR! System Windows_NT 10.0.17134
gyp ERR! command "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\node_modules\\node-gyp\\bin\\node-gyp.js" "rebuild"
gyp ERR! cwd C:\Users\*****\Desktop\node-swipl-master
gyp ERR! node -v v10.15.2
gyp ERR! node-gyp -v v3.8.0
gyp ERR! not ok
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! swipl@1.0.4 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the swipl@1.0.4 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

```

**Comment by rla** [comment]
Which version of SWI did you install? I tested with 8.0.3 yesterday. Some very old version might be missing the constants mentioned in the errors.

**Comment by FerrariAndrea** [comment]
Thx, i have and old version:
```
C:\Users\****>swipl --version
SWI-Prolog version 6.4.0 for x64-win64
```
Now i have installed the 8.0.3 but on "npm install swipl" or "npm install rebuild"  or "npm install node-gyp" i recive the same error:

```
> swipl@1.0.3 install C:\*****\Server\node_modules\swipl
> node-gyp rebuild


C:\*****\Server\node_modules\swipl>if not defined npm_config_node_gyp (node "C:\Program Files\nodejs\node_modules\npm\node_modules\npm-lifecycle\node-gyp-bin\\..\..\node_modules\node-gyp\bin\node-gyp.js" rebuild )  else (node "C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\bin\node-gyp.js" rebuild )
Compilazione dei progetti nella soluzione uno alla volta. Per abilitare la compilazione parallela, aggiungere l'opzione "/m".
  libswipl.cc
  win_delay_load_hook.cc
c:\mydata\compilationhelper\server\node_modules\swipl\src\libswipl.cc(43): warning C4996: 'v8::String::Utf8Value::Utf8V
alue': è stato dichiarato deprecato [C:\*****\Server\node_modules\swipl\build\swipl.vcxproj]
  c:\users\*****\.node-gyp\10.15.2\include\node\v8.h(2891): note: vedere la dichiarazione di 'v8::String::Utf8
  Value::Utf8Value'
c:\mydata\compilationhelper\server\node_modules\swipl\src\libswipl.cc(253): warning C4996: 'v8::String::Utf8Value::Utf8
Value': è stato dichiarato deprecato [C:\*****\Server\node_modules\swipl\build\swipl.vcxproj]
  c:\users\*****\.node-gyp\10.15.2\include\node\v8.h(2891): note: vedere la dichiarazione di 'v8::String::Utf8
  Value::Utf8Value'
LINK : fatal error LNK1181: impossibile aprire il file di input 'c:\program files\swipl\lib\libswipl.lib' [C:\MyData\Co
mpilationHelper\Server\node_modules\swipl\build\swipl.vcxproj]
gyp ERR! build error
gyp ERR! stack Error: `C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\MSBuild\15.0\Bin\MSBuild.exe` failed with exit code: 1
gyp ERR! stack     at ChildProcess.onExit (C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\lib\build.js:262:23)
gyp ERR! stack     at ChildProcess.emit (events.js:189:13)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:248:12)
gyp ERR! System Windows_NT 10.0.17134
gyp ERR! command "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\node_modules\\node-gyp\\bin\\node-gyp.js" "rebuild"
gyp ERR! cwd C:\*****\Server\node_modules\swipl
gyp ERR! node -v v10.15.2
gyp ERR! node-gyp -v v3.8.0
gyp ERR! not ok
npm WARN compilation_helper@1.0.0 No description

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! swipl@1.0.3 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the swipl@1.0.3 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
```

This only if i try to add swipl on my app. If i download or clone your repo and do "npm install" all be good:

```
C:\Users\*****\Desktop\node-swipl-master>npm install
npm WARN deprecated json3@3.3.2: Please use the native JSON object instead of JSON 3

> swipl@1.0.4 install C:\Users\*****\Desktop\node-swipl-master
> node-gyp rebuild


C:\Users\*****\Desktop\node-swipl-master>if not defined npm_config_node_gyp (node "C:\Program Files\nodejs\node_modules\npm\node_modules\npm-lifecycle\node-gyp-bin\\..\..\node_modules\node-gyp\bin\node-gyp.js" rebuild )  else (node "C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\bin\node-gyp.js" rebuild )
Compilazione dei progetti nella soluzione uno alla volta. Per abilitare la compilazione parallela, aggiungere l'opzione "/m".
  libswipl.cc
  win_delay_load_hook.cc
     Creazione della libreria C:\Users\*****\Desktop\node-swipl-master\build\Release\libswipl.lib e dell'ogget
  to C:\Users\*****\Desktop\node-swipl-master\build\Release\libswipl.exp
  Generazione codice in corso...
  All 476 functions were compiled because no usable IPDB/IOBJ from previous compilation was found.
  Generazione codice terminata
  swipl.vcxproj -> C:\Users\*****\Desktop\node-swipl-master\build\Release\\libswipl.node

> swipl@1.0.4 postinstall C:\Users\*****\Desktop\node-swipl-master
> node scripts/store_plbase.js

npm notice created a lockfile as package-lock.json. You should commit this file.
added 122 packages from 358 contributors and audited 193 packages in 12.562s
found 2 vulnerabilities (1 low, 1 critical)
  run `npm audit fix` to fix them, or `npm audit` for details
```

It is a dependency error?
(i run the "npm install --global --production windows-build-tools" and try do copy the "build" dir of your repo to my dir, it does't change the result)


Edit:
I don't have this lib: `c:\program files\swipl\lib\libswipl.lib`
in the path `c:\program files\swipl\lib\` i have only "jpl.jar" if during the SWI-Prolog 8.0.3 installation i agree to remove old version (6.4.0), else if i don'agree in the path i have these file:
`jpl.jar`
`libpthreadGC2.dll.a`
`libswipl.dll.a`
`plterm.dll.a`

**Comment by FerrariAndrea** [comment]
I think this issues can be considered closed.

Now i m using your repo as base for my project, and it seems do work on all my pc.

Anyway I noticed that, the problem maybe is on open cmd as admin for run `npm install --global --production windows-build-tools` because the cmd point to `C:\System32` instead the npm-app path.

it was enough to move through `cd /d npm-app-path`  befoure run `npm install --global --production windows-build-tools` 

Thx for all :)

**Comment by FerrariAndrea** [comment]
Now i m sure, the problem is about the version of SWI-Prolog, i had the 8.x, after downgrade to the 7.x `the npm install swipl` work.

**Comment by rla** [comment]
@FerrariAndrea, please try to install from npm and with 8.x again. I was able to publish fixed version (1.0.4) of the package to the registry but i was not able to test it since my Internet connection is too bad for installing compiler tools on Windows.

**Comment by FerrariAndrea** [comment]
I confirm that it works, nice job.

My actual verion of swipl ($ swipl --version):
```
SWI-Prolog version 8.0.3 for x64-win64
```

create a new npm project and run `npm install swipl`
result:
```
> swipl@1.0.4 install C:\Users\***\Desktop\butta\node_modules\swipl
> node-gyp rebuild


C:\Users\***\Desktop\butta\node_modules\swipl>if not defined npm_config_node_gyp (node "C:\Program Files\nodejs\node_modules\npm\node_modules\npm-lifecycle\node-gyp-bin\\..\..\node_modules\node-gyp\bin\node-gyp.js" rebuild )  else (node "C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\bin\node-gyp.js" rebuild )
Compilazione dei progetti nella soluzione uno alla volta. Per abilitare la compilazione parallela, aggiungere l'opzione "/m".
  libswipl.cc
  win_delay_load_hook.cc
c:\users\***\desktop\butta\node_modules\swipl\src\libswipl.cc(197): warning C4244: 'inizializzazione': conversione da 'term_t' a 'int'. Possibile perdita di dati. [C:\Users\***\Desktop\butta\node_modules\swipl\build\swipl.vcxproj]
     Creazione della libreria C:\Users\***\Desktop\butta\node_modules\swipl\build\Release\libswipl.lib e dell'oggetto C:\Users\***\Desktop\butta\node_modules\swipl\build\Release\libswipl.exp
  Generazione codice in corso...
  Generazione codice terminata
  swipl.vcxproj -> C:\Users\***\Desktop\butta\node_modules\swipl\build\Release\\libswipl.node

> swipl@1.0.4 postinstall C:\Users\***\Desktop\butta\node_modules\swipl
> node scripts/store_plbase.js

npm notice created a lockfile as package-lock.json. You should commit this file.
npm WARN butta@1.0.0 No description
npm WARN butta@1.0.0 No repository field.

+ swipl@1.0.4
added 2 packages in 10.124s
```

Mini node test code:
```
const swipl = require('swipl');
swipl.call("assert(test(1))");
console.log("result is: ",swipl.call("test(X)").X);
```

result:

```
$ node prova.js
result is:  1

```
